### ISSUE TITLE
 Unable to configure rke2-ingress-nginx as a DaemonSet  (UPGRADE FAILED: Invalid Semantic Version)

### ISSUE THREAD

**Issue by dacrystal** [rancher/rke2#493]
## Issue
Due to `tag` format of  `rke2-ingress-nginx` image, configure `ingress-nginx` to be `DeamonSet` using the following, always fails.

**/var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml**
```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
```

Log (helm-install-rke2-ingress-nginx) 
```
Error: UPGRADE FAILED: template: rke2-ingress-nginx/templates/controller-daemonset.yaml:80:22: executing "rke2-ingress-nginx/templates/controller-daemonset.yaml" at <semverCompare ">=0.9.0-beta.1" .Values.controller.image.tag>: error calling semverCompare: Invalid Semantic Version
```

## Using the following work (as workaround):
```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
      image:
        repository: us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller
        tag: "v0.34.1"
```

## Suggestion:
The `tag` format in `rke2-ingress-nginx` should following [Semantic Versioning 2.0.0](https://semver.org/). One fix it to remove `nginx-` prefix from tag format; e.g.  `0.30.0-rancher1`.

Playground(semverCompare): https://play.golang.org/p/8hLyr2zCkKn

**Comment by bryson** [comment]
I've created a detailed bug report regarding this issue on rancher/rancher#31431 since that bug discusses this issue but has more-recent activity.

@dacrystal Unfortunately this workaround does not work for me, as detailed in my other bug report comment
@rancher-max Just pinging you here since you're involved in both the other bug reports here that lead me to the final issue in the rancher/rancher repo where my bug report is posted.

**Comment by bryson** [comment]
Attempting to create the ingress as a `DaemonSet` fails for me in a couple different ways:

> ```yaml
> apiVersion: helm.cattle.io/v1
> kind: HelmChartConfig
> metadata:
>   name: rke2-ingress-nginx
>   namespace: kube-system
> spec:
>   valuesContent: |-
>     controller:
>       kind: DaemonSet
>       daemonset:
>         useHostPort: true
> ```

When Helm attempts to install Rancher 2 using the config quoted above, it fails a semantic version check:

```bash
$ kubectl logs -n kube-system helm-install-rke2-ingress-nginx-v6sl7
...
helm_v3 install --set-string global.rke2DataDir=/var/lib/rancher/rke2 --set-string global.systemDefaultRegistry=index.docker.io rke2-ingress-nginx /rke2-ingress-nginx.tgz --values /config/values-10_HelmChartConfig.yaml
Error: template: rke2-ingress-nginx/templates/controller-daemonset.yaml:80:22: executing "rke2-ingress-nginx/templates/controller-daemonset.yaml" at <semverCompare ">=0.9.0-beta.1" .Values.controller.image.tag>: error calling semverCompare: Invalid Semantic Version
```

I assume this is because the defaults from the chart [here](https://github.com/rancher/rke2-charts/blob/main-source/packages/rke2-ingress-nginx/generated-changes/patch/values.yaml.patch) include `tag: "nginx-0.35.0-rancher2"`.

When using those defaults directly in my `HelmChartConfig`, I get exactly the same error:

```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
      image:
        repository: rancher/nginx-ingress-controller
        tag: "nginx-0.35.0-rancher2"
```
```bash
$ kubectl logs -n kube-system helm-install-rke2-ingress-nginx-pzq7n
...
helm_v3 install --set-string global.rke2DataDir=/var/lib/rancher/rke2 --set-string global.systemDefaultRegistry=index.docker.io rke2-ingress-nginx /rke2-ingress-nginx.tgz --values /config/values-10_HelmChartConfig.yaml
Error: template: rke2-ingress-nginx/templates/controller-daemonset.yaml:80:22: executing "rke2-ingress-nginx/templates/controller-daemonset.yaml" at <semverCompare ">=0.9.0-beta.1" .Values.controller.image.tag>: error calling semverCompare: Invalid Semantic Version
```

Helm fails to install because the tag `nginx-0.35.0-rancher2` does not match semantic versioning standards with the `nginx-` prefix, and Helm refuses to accept the tag outright. As I understand it, unless the tag format of this image is changed in the `index.docker.io` repository itself, this will never work...unless there is an option to disable/bypass that semantic version check.

When changing the `repository` and `tag` values as instructed in the Rancher2 "HA on RKE2" documentation [here](https://rancher.com/docs/rancher/v2.5/en/installation/resources/k8s-tutorials/ha-rke2/) and described as the workaround above the Helm installation completes but the ingress pods fail to load:
```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
      image:
        repository: us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller
        tag: "v0.34.1"
```
```bash
$ kubectl logs -n kube-system rke2-ingress-nginx-controller-5vhx7
Error from server (BadRequest): container "rke2-ingress-nginx-controller" in pod "rke2-ingress-nginx-controller-5vhx7" is waiting to start: trying and failing to pull image
```
```bash
$ kubectl describe pod -n kube-system rke2-ingress-nginx-controller-5vhx7
Name:         rke2-ingress-nginx-controller-5vhx7
Namespace:    kube-system
...
Controlled By:  DaemonSet/rke2-ingress-nginx-controller
Containers:
  rke2-ingress-nginx-controller:
    Container ID:
    Image:         index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1
 ...
Events:
  Type     Reason     Age                     From               Message
  ----     ------     ----                    ----               -------
  Normal   Scheduled  8m18s                   default-scheduler  Successfully assigned kube-system/rke2-ingress-nginx-controller-5vhx7 to ip-10-0-11-118
  Normal   Pulling    6m40s (x4 over 8m17s)   kubelet            Pulling image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1"
  Warning  Failed     6m38s (x4 over 8m16s)   kubelet            Failed to pull image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": failed to resolve reference "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
  Warning  Failed     6m38s (x4 over 8m16s)   kubelet            Error: ErrImagePull
  Normal   BackOff    6m12s (x6 over 8m15s)   kubelet            Back-off pulling image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1"
  Warning  Failed     3m16s (x19 over 8m15s)  kubelet            Error: ImagePullBackOff
```

The problem here is that the image that the containers are trying to pull is _not_ the image from the YAML config, it is that image with the string `index.docker.io/` prepended, as you can see in the events list: 

```
Failed to pull image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": failed to resolve reference "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

I notice in the first failure using your quoted config that Helm's install command includes `--set-string global.systemDefaultRegistry=index.docker.io`. Not sure that this is related as it's just a default, and I wouldn't expect the default to override the value given...

As far as I can tell, there is no "correct" way to create this DaemonSet in an automated way with this combination of bugs. Is there another solution? 

Currently as a work-around I use the YAML config from the Rancher docs so Helm's install "completes" and the Pods are created/scheduled but image pull fails, and I then set the DaemonSet image manually using Terraform's `depends_on` directives to create a bottleneck for everything that happens after the cluster is expected to be ready:

```
resource "null_resource" "kubernetes_cluster_init" {
  provisioner "local-exec" {
    interpreter = ["bash", "-c"]
    command     = "until kubectl set image -n kube-system ds/rke2-ingress-nginx-controller rke2-ingress-nginx-controller=rancher/nginx-ingress-controller:nginx-0.35.0-rancher2; do sleep 5; done"
  }

  depends_on = [
    # ... (RKE2 server node creation, load balancer availability, dns, etc)
    null_resource.get_kubeconfig,
  ]
}

resource "helm_release" "cert_manager" {
  # ...
  depends_on = [null_resource.kubernetes_cluster_init]
}
resource "helm_release" "rancher2" {
  repository = "https://releases.rancher.com/server-charts/latest"
  # ...
  depends_on = [helm_release.cert_manager]
}
```

Helm: `v3.5.3`
Rancher: `v2.5.7`
RKE2: `v1.20.4+rke2r1`
EC2 instances: `ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20201014` (`ami-84556de5` in Gov Cloud)
Terraform:
```
Terraform v0.14.8
+ provider registry.terraform.io/cloudflare/cloudflare v2.19.2
+ provider registry.terraform.io/hashicorp/aws v3.32.0
+ provider registry.terraform.io/hashicorp/helm v2.0.3
+ provider registry.terraform.io/hashicorp/kubernetes v2.0.2
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/rancher/rancher2 v1.12.0
```

**Comment by brandond** [comment]
We fix up the image tags for semver compares in other places, it looks like we might have just missed that one.
https://github.com/rancher/rke2-charts/blob/5b1760854a20d1293764332ce2c79422ea7398f2/charts/rke2-ingress-nginx/rke2-ingress-nginx/3.3.001/templates/_helpers.tpl#L121

**Comment by brandond** [comment]
Need one more pull-through PR before this can be tested

**Comment by ekristen** [comment]
The docs surrounding this makes it impossible to use a custom image if `system_default_registry` is set otherwise it will always prefix the value given to the `repository` value.

**Comment by brandond** [comment]
Yes, the way the Helm charts for Rancher and RKE2 work the system-default-registry is treated as a prefix that is rendered into the templates. For most cases, private registry support is what you will actually want to use.

**Comment by ekristen** [comment]
The only problem with that as soon as you need to change the location or a single image one or the other breaks.

The docs even suggest when moving nginx to a daemonset to use an image from us.gcr.io but that won’t work.

**Comment by brandond** [comment]
Both of the issue you mention are tracked in existing issues and will be fixed in the next release.

**Comment by brandond** [comment]
Both of the issue you mention are tracked in existing issues and will be fixed in the next release.

**Comment by ekristen** [comment]
Nice. Many thanks

**Comment by ShylajaDevadiga** [comment]
Validated fix using commit id c7da840c
rke2 version v1.20.5-dev+c7da840c
rancher version 2.5.7
